sdk/storage/azblob/internal/shared/blob_batch.go

```diff
@@ -0,0 +1,286 @@
+	subReqHeader.WriteString("Content-ID: " + strconv.Itoa(*contentID) + HttpNewline)
+	subReqHeader.WriteString(HttpNewline)
+
+	return subReqHeader.String()
+}
+
+// buildSubRequest is used for building the sub-request. Example:
+// DELETE /container0/blob0 HTTP/1.1
+// x-ms-date: Thu, 14 Jun 2018 16:46:54 GMT
+// Authorization: SharedKey account:G4jjBXA7LI/RnWKIOQ8i9xH4p76pAQ+4Fs4R1VxasaE=
+// Content-Length: 0
+func buildSubRequest(req *policy.Request) string {
+	var batchSubRequest strings.Builder
+	blobPath := req.Raw().URL.Path
+	if len(req.Raw().URL.RawQuery) > 0 {
+		blobPath += "?" + req.Raw().URL.RawQuery
+	}
+
+	batchSubRequest.WriteString(fmt.Sprintf("%s %s %s%s", req.Raw().Method, blobPath, HttpVersion, HttpNewline))
+
+	for k, v := range req.Raw().Header {
+		if strings.EqualFold(k, HeaderXmsVersion) {
+			continue
+		}
+		if len(v) > 0 {
+			batchSubRequest.WriteString(fmt.Sprintf("%v: %v%v", k, v[0], HttpNewline))
+		}
+	}
+
+	batchSubRequest.WriteString(HttpNewline)
+	return batchSubRequest.String()
+}
+
+// CreateBatchRequest creates a new batch request using the sub-requests present in the BlobBatchBuilder.
+//
+// Example of a sub-request in the batch request body:
+//
+//	--batch_357de4f7-6d0b-4e02-8cd2-6361411a9525
+//	Content-Type: application/http
+//	Content-Transfer-Encoding: binary
+//	Content-ID: 0
+//
+//	DELETE /container0/blob0 HTTP/1.1
+//	x-ms-date: Thu, 14 Jun 2018 16:46:54 GMT
+//	Authorization: SharedKey account:G4jjBXA7LI/RnWKIOQ8i9xH4p76pAQ+4Fs4R1VxasaE=
+//	Content-Length: 0
+func CreateBatchRequest(bb *BlobBatchBuilder) (string, string, error) {
+	batchID, err := createBatchID()
+	if err != nil {
+		return "", "", err
+	}
+
+	contentID := 0
+	var batchRequest strings.Builder
+
+	for _, req := range bb.SubRequests {
+		if bb.AuthPolicy != nil {
+			resp, err := bb.AuthPolicy.Do(req)
+			if err != nil && resp != nil {
+				if log.Should(EventSubmitBatch) {
+					log.Writef(EventSubmitBatch, "failed to authorize sub-request for %v.\nError: %v\nResponse status: %v", req.Raw().URL.Path, err.Error(), resp.Status)
+				}
+			}
+		}
+
+		batchRequest.WriteString(createSubReqHeader(&batchID, &contentID))
+		batchRequest.WriteString(buildSubRequest(req))
+		contentID++
+	}
+
+	// add the last line of the request body. It looks like,
+	// --batch_357de4f7-6d0b-4e02-8cd2-6361411a9525--
+	batchRequest.WriteString(getBatchRequestDelimiter(&batchID, true, true) + HttpNewline)
+
+	return batchRequest.String(), batchID, nil
+}
+
+// UpdateSubRequestHeaders updates the sub-request headers.
+// Removes x-ms-version header.
+func UpdateSubRequestHeaders(req *policy.Request) {
+	// remove x-ms-version header from the request header
+	for k := range req.Raw().Header {
+		if strings.EqualFold(k, HeaderXmsVersion) {
+			delete(req.Raw().Header, k)
+		}
+	}
+}
+
+// BlobBatchResponse contains the response for the individual sub-requests.
+type BlobBatchResponse struct {
```

Comment:
It looks like this, and maybe other types, are directly exported via type alias.  In this case, the content should be moved to the `exported` package so it's clear this is part of public surface area.  The `shared` package is for internal-only content that's shared across packages.

Reply:
Will move this to `exported` package.

Reply:
Done